firedrake/interpolation.py

```diff
@@ -1652,99 +1482,100 @@ def _wrap_dummy_mat(self):
-
-        self._sub_interpolators = Isub
-        self.callable = self._assemble_matnest
-
-    def __getitem__(self, item):
-        return self._sub_interpolators[item]
-
-    def __iter__(self):
-        return iter(self._sub_interpolators)
-
-    def _assemble_matnest(self):
-        """Assemble the operator."""
-        shape = tuple(len(a.function_space()) for a in self.arguments)
-        blocks = numpy.full(shape, PETSc.Mat(), dtype=object)
-        # Assemble the sparse block matrix
-        for i in self:
-            blocks[i] = self[i].callable().handle
-        petscmat = PETSc.Mat().createNest(blocks)
-        tensor = firedrake.AssembledMatrix(self.arguments, self.bcs, petscmat)
-        return tensor.M
-
-    def _interpolate(self, *function, output=None, adjoint=False, **kwargs):
-        """Assemble the action."""
-        rank = len(self.arguments)
-        if rank == 0:
-            result = sum(self[i].assemble(**kwargs) for i in self)
-            return output.assign(result) if output else result
-
-        if output is None:
-            output = firedrake.Function(self.arguments[-1].function_space().dual())
-
-        if rank == 1:
-            for k, sub_tensor in enumerate(output.subfunctions):
-                sub_tensor.assign(sum(self[i].assemble(**kwargs) for i in self if i[0] == k))
-        elif rank == 2:
-            for k, sub_tensor in enumerate(output.subfunctions):
-                sub_tensor.assign(sum(self[i]._interpolate(*function, adjoint=adjoint, **kwargs)
-                                      for i in self if i[0] == k))
-        return output
+            Isub[indices] = (get_interpolator(form), sub_bcs)
+
+        return Isub
+
+    def _build_callable(self, tensor=None, bcs=None):
+        Isub = self._get_sub_interpolators(bcs=bcs)
+        V_dest = self.ufl_interpolate.function_space() or self.target_space
+        f = tensor or Function(V_dest)
+        if self.rank == 2:
+            shape = tuple(len(a.function_space()) for a in self.interpolate_args)
+            blocks = numpy.full(shape, PETSc.Mat(), dtype=object)
+            for indices, (interp, sub_bcs) in Isub.items():
+                interp._build_callable(bcs=sub_bcs)
+                blocks[indices] = interp.callable().handle
+            self.handle = PETSc.Mat().createNest(blocks)
+
+            def callable() -> MixedInterpolator:
+                return self
+        elif self.rank == 1:
+            def callable() -> Function | Cofunction:
+                for k, sub_tensor in enumerate(f.subfunctions):
+                    sub_tensor.assign(sum(
+                        interp.assemble(bcs=sub_bcs) for indices, (interp, sub_bcs) in Isub.items() if indices[0] == k
+                    ))
+                return f
+        else:
+            def callable() -> Number:
+                return sum(interp.assemble(bcs=sub_bcs) for (interp, sub_bcs) in Isub.values())
+        self.callable = callable
```

Comment:
This seems wrong. If you call `_build_callable` multiple times with different arguments this will get overwritten.